app/components/base_component.rb

```diff
@@ -0,0 +1,44 @@
+require 'erb'
+
+class BaseComponent
+  include ActiveModel::Validations
+
+  def render_in(view_context, &block)
+    self.class.compile
+    @view_context = view_context
+    @block = block if block
+    validate!
+    rendered_template
+  end
+
+  def self.compile
+    return if @compiled
+    compiler = ERB::Compiler.new('<>')
+    compiler.pre_cmd = ['_erbout=+""']
+    compiler.put_cmd = '_erbout.<<'
+    compiler.insert_cmd = '_erbout.<<'
+    compiler.post_cmd = ['_erbout.respond_to?(:html_safe) ? _erbout.html_safe : _erbout']
+    code, = compiler.compile(File.read(File.join(File.dirname(__FILE__), template_file)))
```

Comment:
this is darker ruby magic than I am familiar with...  where can I look more into what the `_erbout` means?

Reply:
`_erbout` is just a variable, we could name it whatever we'd want. But the `pre_cmd`, etc. dictates how the compiler generates code.

It's an adapted version of: https://www.rubydoc.info/stdlib/erb/ERB/Compiler